s3/backend.go

```diff
@@ -93,51 +92,67 @@ func (s *s3) ListBucket(bucketName string, prefix *gofakes3.Prefix, page gofakes
 	// Workaround for empty prefix
 	prefix.HasPrefix = prefix.Prefix != ""
 
-	// Specify bucket.
-	opts := []api.ObjectsOption{api.ObjectsWithBucket(bucketName)}
-
-	// Handle prefix.
-	var path string // root of bucket
-	if prefix.HasPrefix {
-		if idx := strings.LastIndex(prefix.Prefix, "/"); idx != -1 {
-			path = prefix.Prefix[:idx+1]
-			prefix.Prefix = prefix.Prefix[idx+1:]
-		}
-		opts = append(opts, api.ObjectsWithPrefix(prefix.Prefix))
-	}
-
-	// Handle pagination.
+	// Adjust MaxKeys
 	if page.MaxKeys == 0 {
 		page.MaxKeys = maxKeysDefault
 	}
-	if page.HasMarker {
-		opts = append(opts, api.ObjectsWithMarker(page.Marker))
-		opts = append(opts, api.ObjectsWithLimit(int(page.MaxKeys)))
-	}
 
-	// Fetch all objects of bucket with the given prefix.
-	res, err := s.b.Object(context.Background(), path, opts...)
+	var objects []api.ObjectMetadata
+	response := gofakes3.NewObjectList()
+	if prefix.HasDelimiter {
+		// Handle request with delimiter.
+		opts := []api.ObjectsOption{api.ObjectsWithBucket(bucketName)}
+		if page.HasMarker {
+			opts = append(opts, api.ObjectsWithMarker(page.Marker))
+			opts = append(opts, api.ObjectsWithLimit(int(page.MaxKeys)))
+		}
+		var path string // root of bucket
+		adjustedPrefix := prefix.Prefix
+		if idx := strings.LastIndex(adjustedPrefix, "/"); idx != -1 {
+			path = adjustedPrefix[:idx+1]
+			adjustedPrefix = adjustedPrefix[idx+1:]
+		}
+		if adjustedPrefix != "" {
+			opts = append(opts, api.ObjectsWithPrefix(adjustedPrefix))
+		}
+		var res api.ObjectsResponse
+		res, err = s.b.Object(context.Background(), path, opts...)
+		if err != nil {
+			return nil, gofakes3.ErrorMessage(gofakes3.ErrInternal, err.Error())
+		}
+		objects = res.Entries
+		response.IsTruncated = res.HasMore
+		if response.IsTruncated {
+			response.NextMarker = objects[len(objects)-1].Name
+		}
+	} else {
+		// Handle request without delimiter.
+		var res api.ObjectsListResponse
+		res, err = s.b.ListObjects(context.Background(), bucketName, "/"+prefix.Prefix, page.Marker, int(page.MaxKeys))
```

Comment:
could we handle this elsewhere you think? seems like our store could prepend a `/` to the prefix if it doesn't have one and we'd be fine in all cases too right?

Reply:
I don't want to hide the fact that our object names start with `/` somewhere in the backend. Sure you can just not prepend it like you can for downloads, but this only works for GET requests and seems like hidden knowledge to me. Why not always pass in the object by its actual name?